File: src/main/java/com/mk7a/soulbind/update/UpdateChecker.java
package com.mk7a.soulbind.update;

import org.bukkit.plugin.Plugin;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse.BodyHandlers;
import java.time.Duration;

public final class UpdateChecker {

    private static final URI LATEST_RELEASE = URI.create(
            "https://github.com/mk7a-mc/ItemSoulBind/releases/latest");

    private static final HttpClient CLIENT = HttpClient.newBuilder()
            .followRedirects(HttpClient.Redirect.NEVER)
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    private UpdateChecker() {}

    public static void check(Plugin plugin) {
        String currentTag = "v" + plugin.getDescription().getVersion();
        HttpRequest request = HttpRequest.newBuilder(LATEST_RELEASE)
                .method("HEAD", BodyPublishers.noBody())
                .timeout(Duration.ofSeconds(5))
                .build();

        CLIENT.sendAsync(request, BodyHandlers.discarding())
                .thenApply(r -> r.headers().firstValue("location").orElseThrow())
                .thenApply(loc -> loc.substring(loc.lastIndexOf('/') + 1))
                .thenAccept(latestTag -> {
                    if (!latestTag.equals(currentTag)) {
                        announce(plugin, latestTag);
                    }
                })
                .exceptionally(e -> {
                    plugin.getLogger().warning("Could not check for updates.");
                    return null;
                });
    }

    private static void announce(Plugin plugin, String latestTag) {
        var log = plugin.getLogger();
        log.info("Latest release: " + latestTag + " (running " + plugin.getDescription().getVersion() + ")");
        log.info("Download: https://modrinth.com/plugin/itemsoulbind");
        log.info("Source:   https://github.com/mk7a-mc/ItemSoulBind/releases/latest");
    }
}

File: src/main/kotlin/com/mk7a/soulbind/update/UpdateChecker.kt
package com.mk7a.soulbind.update

import org.bukkit.plugin.Plugin
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/** Resolves the latest release tag from GitHub's `releases/latest` redirect, without touching the API. */
object UpdateChecker {

    private val LATEST_RELEASE = URI.create("https://github.com/mk7a-mc/ItemSoulBind/releases/latest")
    private val TIMEOUT = Duration.ofSeconds(5)

    fun check(plugin: Plugin) {
        val log = plugin.logger
        val currentVersion = plugin.pluginMeta.version

        val client = HttpClient.newBuilder()
            .followRedirects(HttpClient.Redirect.NEVER)
            .connectTimeout(TIMEOUT)
            .build()
        val request = HttpRequest.newBuilder(LATEST_RELEASE)
            .method("HEAD", HttpRequest.BodyPublishers.noBody())
            .timeout(TIMEOUT)
            .build()

        // Completes on the HTTP client's own threads; only logging happens there
        client.sendAsync(request, HttpResponse.BodyHandlers.discarding())
            .thenApply { response -> response.headers().firstValue("location").orElseThrow().substringAfterLast('/') }
            .thenAccept { latestTag ->
                if (latestTag != "v$currentVersion") {
                    log.info("Latest release: $latestTag (running $currentVersion)")
                    log.info("Download: https://modrinth.com/plugin/itemsoulbind")
                    log.info("Source:   $LATEST_RELEASE")
                }
            }
            .exceptionally { log.warning("Could not check for updates."); null }
            .whenComplete { _, _ -> client.shutdown() }
    }
}
